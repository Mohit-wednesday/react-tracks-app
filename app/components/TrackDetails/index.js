/**
 *
 * TrackDetails
 *
 */

import React, { useState, memo, useEffect } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { createStructuredSelector } from 'reselect';
import { injectIntl } from 'react-intl';
import { Helmet } from 'react-helmet';
import { useParams } from 'react-router-dom';
import { injectSaga } from 'redux-injectors';
import { colors, media } from '@app/themes';
import { itunesContainerCreators } from '@app/containers/ItunesContainer/reducer';
import { selectTrackData, selectTrackError } from '@app/containers/ITunesContainer/selectors';
import { connect } from 'react-redux';
import { compose } from 'redux';
import itunesContainerSaga from '@app/containers/ITunesContainer/saga';
import { isEmpty } from 'lodash';
import { Skeleton, Card, Row, Col, Image, Statistic } from 'antd';
import If from '@app/components/If';
import T from '@app/components/T';
import { dateFormat } from '@app/utils/index';

const StyledContainer = styled.div`
  display: flex;
  justify-content: center;
  flex-direction: column;
  margin: 2rem auto;
  max-width: ${(props) => props.maxwidth}px;
  padding: ${(props) => props.padding}rem;
`;

const StyledImage = styled(Image)`
  && {
    height: 100%;
    width: 100%;
    object-fit: cover;
  }
`;

const StyledImageContainer = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  height: 100%;
  width: 100%;
  flex: 0.4;
`;

const TrackContent = styled.div`
  position: relative;
  width: 100%;
  display: flex;
  gap: 1rem;
  align-items: center;
  ${media.lessThan('tablet')`
    flex-direction: column !important;
  `}
`;

const TrackInfo = styled.div`
  flex: 0.6;
`;

const TrackName = styled(T)`
  && {
    font-size: 1.6rem;
    margin-bottom: 0.5rem;
    font-weight: 700;
    color: ${colors.primary};
  }
`;
const ArtistName = styled(T)`
  && {
    margin-top: 0rem;
    font-size: 1.2rem;
    font-weight: 600;
    margin-bottom: 0.6rem;
    opacity: 0.5;
  }
`;

const ReleaseDate = styled(T)`
  && {
    margin-bottom: 0.2rem;
    font-weight: bold;
    color: ${colors.subText};
    width: 100%;
  }
`;

const CollectionName = styled(T)`
  && {
    margin: 0.5rem 0;
  }
`;

const Genre = styled(T)`
  && {
    margin: 0.5rem 0;
    max-width: fit-content;
    padding: 0.1rem 0.5rem;
    border-radius: 99px;
    color: ${colors.trackCardColor};
    background-color: ${colors.primary};
  }
`;

const StyledAudio = styled.audio`
  && {
    margin: 0.5rem 0 0 0;
  }
`;

const valueStyle = { fontSize: '1.2rem', color: colors.subText };

function TrackDetails({ dispatchGetTrackDetails, track, trackError, maxwidth, padding }) {
  const [loading, setLoading] = useState(false);
  const params = useParams();
  const {
    artworkUrl100,
    artistName,
    collectionName,
    primaryGenreName,
    trackName,
    trackPrice,
    previewUrl,
    releaseDate,
    trackTimeMillis,
    country
  } = track;

  useEffect(() => {
    if (!isEmpty(track) || !isEmpty(trackError)) {
      setLoading(false);
    }
    let loadingTimerId = setTimeout(() => {
      setLoading(false);
    }, 3000);
    return () => {
      clearTimeout(loadingTimerId);
    };
  }, [track, trackError]);

  useEffect(() => {
    if (dispatchGetTrackDetails) {
      dispatchGetTrackDetails(params.trackId);
    }
  }, [params.trackId]);

  return (
    <StyledContainer data-testid="track-details" maxwidth={maxwidth} padding={padding}>
      <Helmet>
        <title>
          {track?.trackName
            ? `${track?.trackName}${track?.artistName && ` by ${track?.artistName}`}`
            : 'TrackContainer'}
        </title>
        <meta name="description" content={track?.trackName ?? 'Description of TrackContainer'} />
      </Helmet>
      <Skeleton loading={loading} active data-testid="skeleton-loader">
        <Card>
          <TrackContent>
            <StyledImageContainer>
              <StyledImage
                width="95%"
                height="95%"
                src={artworkUrl100 ?? ''}
                fallback="https://media.istockphoto.com/vectors/thumbnail-image-vector-graphic-vector-id1147544807?k=20&m=1147544807&s=612x612&w=0&h=pBhz1dkwsCMq37Udtp9sfxbjaMl27JUapoyYpQm0anc="
              ></StyledImage>
            </StyledImageContainer>
            <TrackInfo>
              <If
                condition={trackName}
                otherwise={<TrackName id="track_name_unavailable" data-testid="track-name-unavailable" />}
              >
                <TrackName title={trackName} text={trackName} data-testid="track-name" />
              </If>
              <If
                condition={artistName}
                otherwise={<ArtistName data-testid="artist_name_unavailable" id="artist_name_unavailable" />}
              >
                <ArtistName title={artistName} text={artistName} data-testid="track-artist" />
              </If>
              <If
                condition={releaseDate}
                otherwise={<ReleaseDate data-testid="release_date_unavailable" id="release_date_unavailable" />}
              >
                <ReleaseDate
                  title={releaseDate}
                  text={`Release Date: ${dateFormat(releaseDate)}`}
                  data-testid="track-release"
                />
              </If>
              <If
                condition={collectionName}
                otherwise={
                  <CollectionName data-testid="collection_name_unavailable" id="collection_name_unavailable" />
                }
              >
                <CollectionName title={collectionName} text={collectionName} data-testid="track-collection-name" />
              </If>
              <If
                condition={primaryGenreName}
                otherwise={<Genre data-testid="genre_unavailable" id="genre_unavailable" />}
              >
                <Genre title={primaryGenreName} text={primaryGenreName} data-testid="track-genre" />
              </If>
              <Row>
                <Col span={8}>
                  <Statistic
                    data-testid="track-price"
                    suffix="$"
                    precision={2}
                    title="Track Price"
                    value={trackPrice}
                    valueStyle={valueStyle}
                  />
                </Col>
                <Col span={8}>
                  <Statistic
                    data-testid="track-time"
                    precision={2}
                    suffix=" s"
                    title="Track Time"
                    value={trackTimeMillis / 1000}
                    valueStyle={valueStyle}
                  />
                </Col>
                <Col span={8}>
                  <Statistic
                    data-testid="track-country"
                    precision={2}
                    title="Country"
                    value={country}
                    valueStyle={valueStyle}
                  />
                </Col>
              </Row>
              <If
                condition={previewUrl}
                otherwise={<T data-testid="track_preview_unavailable" id="track_preview_unavailable" />}
              >
                <StyledAudio loading="lazy" preload="none" src={previewUrl} controls />
              </If>
            </TrackInfo>
          </TrackContent>
        </Card>
      </Skeleton>
    </StyledContainer>
  );
}

TrackDetails.defaultProps = {
  maxwidth: 800,
  padding: 1
};

TrackDetails.propTypes = {
  dispatchGetTrackDetails: PropTypes.func,
  track: PropTypes.object,
  trackError: PropTypes.any,
  maxwidth: PropTypes.number,
  padding: PropTypes.number
};

export function mapDispatchToProps(dispatch) {
  const { requestGetTrackDetails } = itunesContainerCreators;
  return {
    dispatchGetTrackDetails: (trackId) => dispatch(requestGetTrackDetails(trackId))
  };
}
const mapStateToProps = createStructuredSelector({
  track: selectTrackData(),
  trackError: selectTrackError()
});

const withConnect = connect(mapStateToProps, mapDispatchToProps);

export default compose(
  withConnect,
  memo,
  injectSaga({ key: 'itunesContainer', saga: itunesContainerSaga })
)(TrackDetails);

export const TrackContainerTest = compose(injectIntl)(TrackDetails);
