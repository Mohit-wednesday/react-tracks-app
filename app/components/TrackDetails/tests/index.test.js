/**
 *
 * Tests for TrackDetails
 *
 */

import React from 'react';
// import { fireEvent } from '@testing-library/dom'
import { renderProvider } from '@utils/testUtils';
import TrackDetails from '../index';

describe('<TrackDetails />', () => {
  it('should render and match the snapshot', () => {
    const { baseElement } = renderProvider(<TrackDetails />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should contain 1 TrackDetails component', () => {
    const { getAllByTestId } = renderProvider(<TrackDetails />);
    expect(getAllByTestId('track-details').length).toBe(1);
  });
});
