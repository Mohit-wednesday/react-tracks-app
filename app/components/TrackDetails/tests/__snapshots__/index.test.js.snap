// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TrackDetails /> should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="TrackDetails__StyledContainer-sc-18wb0c5-0 bfxYEi"
      data-testid="track-details"
    >
      <div
        class="ant-card ant-card-bordered"
      >
        <div
          class="ant-card-body"
        >
          <div
            class="TrackDetails__TrackContent-sc-18wb0c5-3 cwZvTp"
          >
            <div
              class="TrackDetails__StyledImageContainer-sc-18wb0c5-2 doYUpH"
            >
              <div
                class="ant-image"
                style="width: 95%; height: 95%;"
              >
                <img
                  class="ant-image-img TrackDetails__StyledImage-sc-18wb0c5-1 gzUmxq"
                  src=""
                  style="height: 95%;"
                />
                <div
                  class="ant-image-mask"
                >
                  <div
                    class="ant-image-mask-info"
                  >
                    <span
                      aria-label="eye"
                      class="anticon anticon-eye"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="eye"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M942.2 486.2C847.4 286.5 704.1 186 512 186c-192.2 0-335.4 100.5-430.2 300.3a60.3 60.3 0 000 51.5C176.6 737.5 319.9 838 512 838c192.2 0 335.4-100.5 430.2-300.3 7.7-16.2 7.7-35 0-51.5zM512 766c-161.3 0-279.4-81.8-362.7-254C232.6 339.8 350.7 258 512 258c161.3 0 279.4 81.8 362.7 254C791.5 684.2 673.4 766 512 766zm-4-430c-97.2 0-176 78.8-176 176s78.8 176 176 176 176-78.8 176-176-78.8-176-176-176zm0 288c-61.9 0-112-50.1-112-112s50.1-112 112-112 112 50.1 112 112-50.1 112-112 112z"
                        />
                      </svg>
                    </span>
                    Preview
                  </div>
                </div>
              </div>
            </div>
            <div
              class="TrackDetails__TrackInfo-sc-18wb0c5-4 iojZno"
            >
              <p
                class="T__StyledText-gjlic1-0 jbStwN TrackDetails__TrackName-sc-18wb0c5-5 cwWSZm"
                data-testid="track-name-unavailable"
              >
                Track name is unavailable
              </p>
              <p
                class="T__StyledText-gjlic1-0 jbStwN TrackDetails__ArtistName-sc-18wb0c5-6 jrVmVq"
                data-testid="artist_name_unavailable"
              >
                Artist name is unavailable
              </p>
              <p
                class="T__StyledText-gjlic1-0 jbStwN TrackDetails__ReleaseDate-sc-18wb0c5-7 ioqxwl"
                data-testid="release_date_unavailable"
              >
                Release date is unavailable
              </p>
              <p
                class="T__StyledText-gjlic1-0 jbStwN TrackDetails__CollectionName-sc-18wb0c5-8 gKrRoq"
                data-testid="collection_name_unavailable"
              >
                Collection name is unavailable
              </p>
              <p
                class="T__StyledText-gjlic1-0 jbStwN TrackDetails__Genre-sc-18wb0c5-9 gIZbJH"
                data-testid="genre_unavailable"
              >
                genre is unavailable
              </p>
              <div
                class="ant-row"
              >
                <div
                  class="ant-col ant-col-8"
                >
                  <div
                    class="ant-statistic"
                  >
                    <div
                      class="ant-statistic-title"
                    >
                      Track Price
                    </div>
                    <div
                      class="ant-statistic-content"
                      style="font-size: 1.2rem; color: rgb(128, 128, 128);"
                    >
                      <span
                        class="ant-statistic-content-value"
                      >
                        <span
                          class="ant-statistic-content-value-int"
                        >
                          
                          0
                        </span>
                        <span
                          class="ant-statistic-content-value-decimal"
                        >
                          .00
                        </span>
                      </span>
                      <span
                        class="ant-statistic-content-suffix"
                      >
                        $
                      </span>
                    </div>
                  </div>
                </div>
                <div
                  class="ant-col ant-col-8"
                >
                  <div
                    class="ant-statistic"
                  >
                    <div
                      class="ant-statistic-title"
                    >
                      Track Time
                    </div>
                    <div
                      class="ant-statistic-content"
                      style="font-size: 1.2rem; color: rgb(128, 128, 128);"
                    >
                      <span
                        class="ant-statistic-content-value"
                      >
                        NaN
                      </span>
                      <span
                        class="ant-statistic-content-suffix"
                      >
                         s
                      </span>
                    </div>
                  </div>
                </div>
                <div
                  class="ant-col ant-col-8"
                >
                  <div
                    class="ant-statistic"
                  >
                    <div
                      class="ant-statistic-title"
                    >
                      Country
                    </div>
                    <div
                      class="ant-statistic-content"
                      style="font-size: 1.2rem; color: rgb(128, 128, 128);"
                    >
                      <span
                        class="ant-statistic-content-value"
                      >
                        <span
                          class="ant-statistic-content-value-int"
                        >
                          
                          0
                        </span>
                        <span
                          class="ant-statistic-content-value-decimal"
                        >
                          .00
                        </span>
                      </span>
                    </div>
                  </div>
                </div>
              </div>
              <p
                class="T__StyledText-gjlic1-0 jbStwN"
                data-testid="track_preview_unavailable"
              >
                Track preview is unavailable
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
